tapgarden/re-org_watcher.go

```diff
@@ -0,0 +1,424 @@
+// WatchProofs adds new proofs to the re-org watcher for their anchor
+// transaction to be watched until it reaches a safe confirmation depth.
+func (w *ReOrgWatcher) WatchProofs(newProofs []*proof.Proof,
+	onProofUpdate ProofUpdateCb) {
+
+	if len(newProofs) == 0 {
+		return
+	}
+
+	select {
+	case w.incomingProofs <- &proofRegistration{
+		proofs:   newProofs,
+		updateCb: onProofUpdate,
+	}:
+	case <-w.Quit:
+	}
+}
+
+// reportErr reports an error to the main server.
+func (w *ReOrgWatcher) reportErr(err error) {
+	select {
+	case w.cfg.ErrChan <- err:
+	case <-w.Quit:
+	}
+}
+
+// waitForConf waits for the anchor transaction of the given proofs to reach a
+// safe confirmation depth.
+func (w *ReOrgWatcher) waitForConf(ctx context.Context,
+	newProofs *proofRegistration) error {
+
+	firstProof := newProofs.proofs[0]
+	tx := firstProof.AnchorTx
+	txHash := tx.TxHash()
+
+	// Do we already have a confirmation watcher for this transaction? Then
+	// we don't need to add another one.
+	existingNtfn, ok := w.pendingProofs[txHash]
+	if ok {
+		existingNtfn.proofsRegistrations = append(
+			existingNtfn.proofsRegistrations, newProofs,
+		)
+
+		return nil
+	}
+
+	reOrgChan := make(chan struct{}, 1)
+	ctxc, cancel := context.WithCancel(ctx)
+
+	// We only register for a single confirmation, which we expect to come
+	// in immediately, since we're only given the proof after one conf.
+	// But we keep the registration open, so we will get notified again if
+	// there is a re-org.
+	confEvent, errChan, err := w.cfg.ChainBridge.RegisterConfirmationsNtfn(
+		ctxc, &txHash, tx.TxOut[0].PkScript, 1, firstProof.BlockHeight,
+		true, reOrgChan,
+	)
+	if err != nil {
+		cancel()
+		return fmt.Errorf("unable to register for conf ntfn: %w", err)
+	}
+
+	// Now that we have created the confirmation watcher, we can add the
+	// new proofs to the pending proofs map.
+	w.pendingProofs[txHash] = &anchorTxNotification{
+		proofsRegistrations: []*proofRegistration{newProofs},
+		cancel:              cancel,
+	}
+
+	w.Wg.Add(1)
+	go func() {
+		defer confEvent.Cancel()
+		defer w.Wg.Done()
+		defer cancel()
+
+		for {
+			select {
+			// If the transaction was confirmed normally, we don't
+			// need to do anything and can stop watching it.
+			case conf, ok := <-confEvent.Confirmed:
+				if !ok {
+					return
+				}
+
+				select {
+				case w.incomingConfs <- conf:
+				case <-w.Quit:
+				}
+
+			// If the transaction was re-organized out, we need to
+			// update the proof and store it in the proof archive.
+			case <-reOrgChan:
+				log.Infof("Anchor TX %v was re-organized out "+
+					"of the chain, will update proof "+
+					"with next confirmation", txHash)
+
+				// We continue to watch the transaction until
+				// it reaches a safe confirmation depth. We
+				// basically only use this signal to log the
+				// message above. We expect another confirmation
+				// to come in once the transaction is included
+				// in a new block in the re-organized chain.
```

Comment:
Would think we also need to send a signal back to the main goroutine that a re-org occurred.

Reply:
We can't really do anything when the TX is re-organized out of the chain. That basically meant it went back into the mempool. So we just log the info here and wait for it to be confirmed in a block again.
Or do you suggest we mark the proof as invalid here as it's possible it will never make it into the chain again?

Reply:
It seems to me that informing the main goroutine might come in handy when it comes to signalling to a user (via UI) that the previously confirmed send event is now unconfirmed again.

Reply:
Re-orgs happen very rarely AFAIK. And I would assume that if they happen the new chain would very likely contain very similar transactions. So the chance of a transaction being un-confirmed for a long time is probably quite low. So I'm not sure we actually want to build out any user-facing thing here. And if we do, I'd suggest doing it in a follow-up PR.

Reply:
> Or do you suggest we mark the proof as invalid here as it's possible it will never make it into the chain again?

I think I was just sort of confused re the control flow here (message passing ordering). Will take another pass over things with fresh eyes.